### PR TITLE
feat(ext-plugin): support to get request body

### PR DESCRIPTION
**Pull request by tzssangglass**
Signed-off-by: tzssangglass <tzssangglass@gmail.com>

### Pre-submission checklist:

* [x] Did you explain what problem does this PR solve? Or what new features have been added?
* [x] Have you added corresponding test cases?
* [ ] Have you modified the corresponding document?
* [x] Is this PR backward compatible? **If it is not backward compatible, please discuss on the [mailing list](https://github.com/apache/apisix/tree/master#community) first**

### CODE REVIEW

**Comment by tzssangglass** [comment]
I'm thinking that this implementation might not be good enough, because usually the user needs to not only read the request body, but also rewrite it.

File: t/lib/ext-plugin.lua [hunk 1]
```diff
@@ -169,6 +169,8 @@ function _M.go(case)
             local entry = call_req:Args(1)
             assert(entry:Name() == "x")
             assert(entry:Value() == "z")
+        elseif case.get_request_body then
+            assert(call_req:Method() == a6_method.POST)
```
Comment (spacewander) [inline t/lib/ext-plugin.lua:173]: Do we need to check the request_body's value?

File: t/plugin/ext-plugin/request-body.t [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+    }
+--- error_code: 405
+--- grep_error_log eval
+qr/send extra info req successfully/
+--- grep_error_log_out
+send extra info req successfully
+
+
+
+=== TEST 4: request body(json)
```
Comment (spacewander) [inline t/plugin/ext-plugin/request-body.t:155]: Better to add a test which doesn't have a value

Comment (tzssangglass) [reply t/plugin/ext-plugin/request-body.t:155]: I have tried to emulate this test case：https://github.com/apache/apisix/blob/4b94b84f92d1cb6cb58e52075637a42ed76457d8/t/core/request.t#L414-L439 but failed.

should I send an empty body to test?

File: t/lib/ext-plugin.lua (hunk 1 above)
Comment (tzssangglass) [reply t/lib/ext-plugin.lua:173]: I want to validate different content-type, so the value of request body is not fixed.

Comment (tzssangglass) [reply t/lib/ext-plugin.lua:173]: actions.result would check the value in request body?

File: t/plugin/ext-plugin/request-body.t (hunk 2 above)
Comment (tzssangglass) [reply t/plugin/ext-plugin/request-body.t:155]: add a test case of request body is nil

**Comment by tzssangglass** [comment]
dealy to merge, is: https://github.com/api7/ext-plugin-proto/blob/main/lua/A6/ExtraInfo/ReqBody.lua used for request body?

**Comment by tzssangglass** [comment]
> dealy to merge, is: [api7/ext-plugin-proto@`main`/lua/A6/ExtraInfo/ReqBody.lua](https://github.com/api7/ext-plugin-proto/blob/main/lua/A6/ExtraInfo/ReqBody.lua?rgh-link-date=2021-11-25T08%3A25%3A29Z) used for request body?

ok, after review:https://lists.apache.org/thread/m5g44gtywx05x4s262rn7qg87oxscy0v, the ReqBody in ExtraInfo is used to  holds the request body.

**Comment by tzssangglass** [comment]
I will update this PR.

File: t/plugin/ext-plugin/request-body.t (hunk 2 above)
Comment (spacewander) [reply t/plugin/ext-plugin/request-body.t:155]: Yes

File: apisix/plugins/ext-plugin/init.lua [hunk 3]
```diff
@@ -273,6 +274,12 @@ local function handle_extra_info(ctx, input)
 
         local var_name = var_req:Name()
         res = ctx.var[var_name]
+    elseif info_type == extra_info.ReqBody then
+        local info = req:Info()
+        local reqbody_req = extra_info_reqbody.New()
+        reqbody_req:Init(info.bytes, info.pos)
+
+        res = core.request.get_body()
```
Comment (spacewander) [inline apisix/plugins/ext-plugin/init.lua:282]: Better to log down err if it is not nil.

Comment (tzssangglass) [reply apisix/plugins/ext-plugin/init.lua]: done

File: apisix/plugins/ext-plugin/init.lua [hunk 4]
```diff
@@ -279,7 +279,12 @@ local function handle_extra_info(ctx, input)
         local reqbody_req = extra_info_reqbody.New()
         reqbody_req:Init(info.bytes, info.pos)
 
-        res = core.request.get_body()
+        local err
+        res, err = core.request.get_body()
+        if not res then
```
Comment (spacewander) [inline apisix/plugins/ext-plugin/init.lua:284]: We need to check the err instead, as the body can be nil when there is no body.

Comment (tzssangglass) [reply apisix/plugins/ext-plugin/init.lua]: update

**Review by spacewander — APPROVED** [review]

**Review by shuaijinchao — APPROVED** [review]
